rest_specification/restful_service_specification.md

```diff
@@ -0,0 +1,243 @@
+shall be of the form `sort={field}(,{direction})` where `direction` is `asc` or
+`desc`.
+
+### Responses
+
+The response shall be a JSON object formatted according to Spring Data REST
+conventions.Specifically, the list is returned as a pluralized field name
+under `_embedded`, and a `page` object supplies paging info. Example:
+
+```json
+GET /user
+
+{
+  "_embedded": {
+    "users": [
+      {
+        ...
+      },
+      ...
+    ]
+  },
+  "page": {
+    "size": 20,
+    "totalElements": 200,
+    "totalPages": 10,
+    "number": 0
+  }
+}
+```
+
+## Additional Functions
+
+Additional APIs may expose other functionality related to a resource type.
+Examples include:
+ - query for a specific subset of resources
+ - perform an action against a set of resources
+
+Additional functions shall be `POST` only, and parameters will be provided in
+the request body. The request body SHOULD be a JSON object with fields
+corresponding to parameter names.
+Examples:
+
+```json
+POST /user/findByLogin
+
+{
+  "login": "bob"
+}
+```
+
+Each function that can be performed against a specific resource must exist in
+two forms:
+ - non-batch
+ - batch
+
+The non-batch form shall take any relevant parameters in a JSON object. Example:
+
+```json
+POST /user/123/deactivate
+
+{
+  "reason": "Banned"
+}
+```
+
+The batch form shall take a JSON array of objects, with each object having the
+identifier as an additional field. Example:
+
+```json
+POST /user/deactivate
+
+[
+  {
+    "id": 123,
+    "reason": "Banned"
+  },
+  {
+    "id": 456,
+    "reason": "Banned"
+  }
+]
+```
+
+Responses for functions shall follow [JSend](http://labs.omniti.com/labs/jsend)
+conventions:
+ - Responses must be JSON objects
+ - `status` field shall tell whether the function call was successful
+  - possible values: `success`, `fail`, `error`
+ - `data` shall have any feedback from the function call (function-specific)
+ - `message` shall contain a descriptive error message in the case of an error.
+
+Success example:
+```json
+{
+  "status" : "success",
+  "data" : null
+}
+```
+
+Related resources may be exposed under additional endpoints off a 
+resource-specific endpoint. Example:
+
+    /user/{id}/posts{&fields,page,size,sort}
+
+## Future/Optional Features
+
+In general, clients must ignore any fields that they don't expect. This allows
+backward-compatibility by additional fields.
+
+### Discoverability (HATEOAS)
+
+Per HAL/Spring Data REST conventions, an implementation may enable usage by
+having relationships between entities and APIs exposed via `_links` fields.
+
+Example:
+```json
+{
+  "_links": {
+    "self": {
+      "href": "https://example.com/rest/v1/user{&sort,page,size}",
+      "templated": true
+    },
+    "next": {
+      "href": "https://example.com/rest/v1/user?page=1&size=20{&sort}",
+      "templated": true
+    }
+  },
+  "_embedded": {
+    "users": [
+      {
+        ...
+      },
+      ...
+    ]
+  },
+  "page": {
+    "size": 20,
```

Comment:
It's worth pointing out that this page section is not part of HAL, but just part of Spring Data REST's `Page` section of the response.  I only bring this up because the operations behind finding this information can be expensive depending on the persistence implementation.  For example, a `count *` in PostgreSQL is *really* slow, which is effectively what gets executed on each page to get this info.  I'm a larger fan of adhering to something more like Spring's `Slice` object, which only exposes whether or not there are more pages, with prev / next links.